node/node.go

```diff
@@ -0,0 +1,242 @@
+	}
+
+	return nil, nil
+
+}
+
+type Node struct {
+	resource.Named
+	resource.AlwaysRebuild
+	logger logging.Logger
+
+	nwkSKey []byte
+	AppSKey []byte
+	AppKey  []byte
+
+	Addr   []byte
+	DevEui []byte
+
+	DecoderPath string
+	NodeName    string
+	gateway     sensor.Sensor
+}
+
+func newNode(
+	ctx context.Context,
+	deps resource.Dependencies,
+	conf resource.Config,
+	logger logging.Logger,
+) (sensor.Sensor, error) {
+	cfg, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return nil, err
+	}
+
+	n := &Node{
+		Named:       conf.ResourceName().AsNamed(),
+		logger:      logger,
+		DecoderPath: cfg.DecoderPath,
+		NodeName:    conf.ResourceName().AsNamed().Name().Name,
+	}
+
+	switch cfg.JoinType {
+	case "OTAA", "":
+		appKey, err := hex.DecodeString(cfg.AppKey)
+		if err != nil {
+			return nil, err
+		}
+		n.AppKey = appKey
+
+		devEui, err := hex.DecodeString(cfg.DevEUI)
+		if err != nil {
+			return nil, err
+		}
+		n.DevEui = devEui
+	case "ABP":
+		devAddr, err := hex.DecodeString(cfg.DevAddr)
+		if err != nil {
+			return nil, err
+		}
+
+		n.Addr = devAddr
+
+		appSKey, err := hex.DecodeString(cfg.AppSKey)
+		if err != nil {
+			return nil, err
+		}
+
+		n.AppSKey = appSKey
+	}
+
+	gateway, err := getGateway(ctx, deps)
+	if err != nil {
+		return nil, err
+	}
+
+	// send the device to the gateway.
+	cmd := make(map[string]interface{})
+	cmd["register_device"] = n
+
+	_, err = gateway.DoCommand(ctx, cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	n.gateway = gateway
+
+	return n, nil
+}
+
+// getGateway sends the validate docommand to the gateway to confirm the dependency.
+func getGateway(ctx context.Context, deps resource.Dependencies) (sensor.Sensor, error) {
+	if len(deps) == 0 {
+		return nil, errors.New("must add sx1302-gateway as dependency")
+	}
+	var dep resource.Resource
+
+	// Assuming there's only one dep.
+	for _, val := range deps {
+		dep = val
+	}
+
+	gateway, ok := dep.(sensor.Sensor)
+	if !ok {
+		return nil, errors.New("dependency must be the sx1302-gateway")
+	}
+
+	cmd := make(map[string]interface{})
+	cmd["validate"] = 1
+
+	// Validate that the dependency is the gateway - gateway will return 1.
+	ret, err := gateway.DoCommand(ctx, cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	retVal, ok := ret["validate"]
+	if !ok {
+		return nil, errors.New("dependency must be the sx1302-gateway")
+	}
+	if retVal.(float64) != 1 {
+		return nil, errors.New("dependency must be the sx1302-gateway")
+	}
+	return gateway, nil
+}
+
+func (n *Node) Close(ctx context.Context) error {
+	return nil
+}
+
+func (n *Node) Readings(ctx context.Context, extra map[string]interface{}) (map[string]interface{}, error) {
+	if n.gateway != nil {
+		allReadings, err := n.gateway.Readings(ctx, nil)
+		if err != nil {
+			return map[string]interface{}{}, err
+		}
+
+		reading, ok := allReadings[n.NodeName]
+		if !ok {
+			// no readings availiable yet
+			return map[string]interface{}{}, data.ErrNoCaptureToStore
```

Comment:
This will happen if the lorawan sensor hasn't sent any data yet. Returning this error will prevent an error from spamming in the logs if data capture is on or you're on the control page. However, the error is `no capture from filter module` which does not reflect what's actually happening and may confuse users. Would like reviewers thoughts on what the behavior should be here.

Reply:
If we add a log line before it, does that clarify things, can we join this error with another more descriptive error and get the same behaviour? I'm assuming no for the second question.

Another avenue would be to add a new error to data capture for the same behaviour, that is more descriptive that is suitable for the LoRaWan case.

Reply:
We could wrap the error and display this: `ConnectError: [unknown] no readings available yet: rpc error: code = FailedPrecondition desc = no capture from filter module`. Wouldn't want to add a log line because if data capture/control page is calling readings frequently the log will spam. Open to the idea of adding a new error specific to lorawan, it would make the error more clear.

Reply:
Did wrapping the error work and keep the behaviour where nothing was synced if no data? That's kinda cool if so.

Reply:
yes! tested it out and `errors.Is` works with wrapped error